### PR TITLE
Give every GC shadow root the lifetime of its value, and root a call's result where it lands

### PR DESCRIPTION
**Pull request by aallan**
> **Sequencing: merge AFTER the #1382 fix**, rebased on it. #1382 is a pre-existing collector bug (a false mark candidate writing into a live array's padding) that this branch does not touch; one conformance program, `ch09_nested_builtin_index`, is red in the new eager-GC lane until it lands. Every other gate is green here.

## What

A GC shadow root lived for the **frame**, not for its **value** — and for a host call's result, it did not exist at all. This makes both ends of the rule one rule.

## Root cause

**Roots outlived their values.** Every root a lowering made was reclaimed by exactly one thing: the function epilogue's `$gc_sp` restore, which runs once, on the way out. A frame therefore held a root for every heap pointer its body produced in passing, live or not — `string_length(string_concat(s, "!"))` repeated K times cost K permanent roots for K values that died immediately. With the shadow stack at 4 096 roots, the depth a recursion could reach became a function of how much its body allocated on the way past, and overflow is a bare `unreachable`: no diagnostic, no location, on a program `check` and `verify` both pass.

[#1322](https://github.com/aallan/vera/issues/1322) was this defect in its match-shaped form. The discipline is now general:

- an **expression's** roots live exactly as long as it is forming its value;
- a **statement's** live exactly as long as the statement, after which what survives is what the statement **bound** — read off the environment delta, so a binding is rooted once by one rule whether it came from a `let`, a pair-`let` or a `let`-destructure. Each producer used to root its own, and a scoped restore would then have had to know, per producer, what to put back.

`_scope_match_shadow_roots` is deleted: a match is now an ordinary case of the general rule rather than a special case beside it.

**Roots did not exist at landing.** A call's result arrives on the operand stack, where the conservative scan cannot see it, so whatever is evaluated next may allocate over it. A Vera function's epilogue re-roots its return into the caller's shadow stack; a **host import has no epilogue**. In the emitted WAT the gap is literal — `call $vera.decimal_from_string` followed directly by the sibling argument's `call $alloc`, with no push between. Whether the callee already rooted it is not something the call site can know, and that ignorance was the bug: rooting at the landing site is the rule that does not have to ask. The redundant root a Vera callee's epilogue already planted costs one slot, which the enclosing scope reclaims; the missing root costs the value.

Four Decimal builtins were affected — `decimal_from_string` and `decimal_div` (`Option<Decimal>` pointer), `decimal_compare` (`Ordering` pointer), and `decimal_to_string`, whose result is a **(ptr, len) pair**. The pair is why the rule reads its result shape from `_stack_shape_of` rather than testing for "one i32": a scalar-keyed rule roots three of the four and silently misses the String.

Three consultor disagreements surfaced on the way and were closed at the root, not at their symptoms:

- an ability operation's result **width** was read from the flat function table by membership, so `show(42)` in a file that merely *imports* a module declaring `fn show(@Int -> @Int)` typed a String result as `i64` while the lowering, narrowed by #1299, correctly reached the operation;
- a `handle` expression's stack shape was asked **after** its effect-op registries were torn down, so a nested `handle[State<Nat>]` answered with the enclosing handler's `get`. The lowering now records the shape at the one moment it is computable;
- the five `show`/`hash` push sites of #1376.

## Evidence — red first

**Roots per frame across a K axis**, in every position a pointer-producing call can occupy. Roots per frame rather than trap depth, because a frame holding *r* roots traps at 4 096 / *r* levels, so a raw depth also carries the outermost frame's own constant and would read that constant as slope.

| position | pre-#1322 | this PR's base | head |
|---|---|---|---|
| block statements | 2 / 3 / 5 | 2 / 3 / 5 | **1 / 1 / 1** |
| `let` right-hand sides | 2 / 3 / 5 | 2 / 3 / 5 | **1 / 1 / 1** |
| call arguments | 3 / 5 / 9 | 3 / 5 / 9 | **1 / 1 / 1** |
| match arm bodies *(control)* | 4 / 7 / 13 | 1 / 1 / 1 | 1 / 1 / 1 |
| `where`-helper call *(control)* | 1 / 1 / 1 | 1 / 1 / 1 | 1 / 1 / 1 |
| handler clause body *(control)* | 1 / 1 / 1 | 1 / 1 / 1 | 1 / 1 / 1 |

Twelve cells red at base, green at head. The three controls matter: arm bodies are #1322's own family, already flat, because a general discipline must not undo the special case it subsumes; a helper call and a handler clause were already scoped by the callee's epilogue, so a family that *moved* there would mean the fix had reached something unbroken.

**Pad sweeps, never single layouts.** #1382 showed a single-layout observation can invert its own meaning. Over data-section pad lengths 0..95, at base and head:

| cell | base failing | head failing |
|---|---|---|
| `decimal_from_string` (`Option` ptr) | **96 / 96** | 0 / 96 |
| `decimal_div` (`Option` ptr) | **96 / 96** | 0 / 96 |
| `decimal_to_string` (String **pair**) | **96 / 96** | 0 / 96 |
| `decimal_compare` (`Ordering` ptr) | 0 / 96 | 0 / 96 |

`decimal_compare` is named by the audit and covered by the rule, but no cell reaching it could be built — its `Ordering` arms carry no heap payload to lose. It stays a control, not a claim. An earlier `decimal_to_string` cell passed only because it read the **length**, which survives a use-after-free that corrupts the bytes; the cells now assert on content, with same-sized churn allocations forcing block reuse (base reads back `'\x00\x00\x00\x005…'` for `'12345…'`).

**Retiring `_emit_root_result` is mutation-validated**, because the existing suites do not discriminate: with `_root_landed_value` disabled and the six per-site pushes removed, `test_codegen_gc_rooting`, `test_codegen_gc_reclamation`, `test_codegen_gc_alloc` and `test_db_marshalling` all stay green (123 tests) while three of the new Map cells die with a **bus error** — an unrooted wrapper corrupting memory outright.

**The #1379 programs**: both red at base, green at head. The issue placed the root cause with discovery re-inference; measurement puts it at the unrooted landing site instead, and the PR records that correction.

## Blast radius

- Suite **12 161 passed**, 173 skipped, 26 deselected. Conformance 244/244; examples 43/43 (`check` + `verify`) and all runnable trap-free, **including under `VERA_EAGER_GC=1`**. Browser parity 413.
- mypy clean; `ruff` and `ruff --select S` clean; doc-counts, limitations-sync (both modes), site-assets, diagnostic-fields, explicit-encoding, changelog gate, corpus-canonical, gitleaks — green.
- **Corpus differential vs `release/v0.2.0`: 253 compared, 143 movers, all 143 accounted for by instruction multiset with local indices normalised away — nothing is removed anywhere, and every net addition is `$gc_sp` save / restore / re-root vocabulary.** (A line-level diff over-reports here: it reports a moved-but-identical line as a `-`/`+` pair.)
- Re-baselined pins, each with its accounting: `TestPointerNessResolvesThroughTheAliasChain`'s push counts go 4/4/3 → 5/5/4, which is a **fall** in cost, not a rise — one extra push against two extra restores, so live roots at the deepest point go 3 → 2, 3 → 2, 2 → 1. #1322's pointer-result residual cell flips from `1 + k` to flat, as its own docstring anticipated.

## The new gate

This PR adds the `VERA_EAGER_GC=1` lane — conformance plus examples-run — as a CI job and a pre-commit hook. The knob turns a rooting bug from a timing accident into a deterministic failure, and both #1379 and #1382 were green in the ordinary run and failing under it. D1b's rule extension is what makes the lane green apart from #1382.

Fixes #1371
Fixes #1376
Fixes #1379

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved garbage-collection handling to preserve live values during nested expressions, blocks, matches, recursive calls, and host operations.
  - Improved reliability for Decimal, Map, and Set results returned from host functions.
  - Fixed edge cases affecting `show`, `hash`, arrays, handlers, and pointer-valued results under eager garbage collection.

- **Tests**
  - Added extensive regression coverage, including eager-GC conformance checks and runnable examples.

- **Documentation**
  - Updated testing statistics, hook documentation, project metrics, and changelog entries.

### CODE REVIEW

**Comment by aallan** [comment]
**Adversarial review at `ebd5f299` (fresh worktrees, import pin asserted; every GC claim a 96-pad data-section sweep under `VERA_EAGER_GC=1`, never a single layout). Verdict: FIX ROUND on evidence; no behavioural finding.**

**The rule holds.** Six attacks on the environment-delta half — a binding shadowed inside a nested block with the outer used after; a `let` whose RHS is a match whose arms bind; a where-helper result across an allocation; a closure capturing a heap value across an allocating sibling; a recursion re-binding a heap value at every level; a handler clause body binding across allocations — produce byte-identical values at base and head at every pad. The `decimal_compare` cell the body says could not be constructed was built (`let @Ordering = decimal_compare(...)`, an allocating statement, then the match): 0/96 at base and at head, so labelling it a control is correct. A Map result passed straight into another allocating Map op, and a discarded host-call result, are 0/96 at both. A nested-host-call shape, `decimal_to_string(option_unwrap_or(decimal_div(...)))` across an allocating sibling, is 96/96 red at base and 0/96 at head — a genuinely new cell.

**MEDIUM — the landing-site half is pinned only by cells that crash.** With `_root_landed_value` disabled, the 123 GC/marshalling tests stay green as claimed, `decimal-from-string-1379` reds by assertion, but `decimal-div-1379` and `decimal-to-string-pair` stay green at all 96 pads — the statement/binding half already roots them, and they were red at base only because base had neither half. The three Map cells fail by SIGBUS, which under `pytest -n auto` kills a worker rather than reporting a failure. The nested-host-call shape is also covered by the binding half under the mutation (0/96). So the retirement of the six `_emit_root_result` pushes rests on three crash-mode cells. Remedy: a structural WAT pin (every heap-returning host call immediately followed by the landing push, the set derived from the registry), the crash-prone cells executed in a subprocess so a trap is a reported failure, and docstrings stating which cells discriminate which half.

**MEDIUM — the pad sweeps are not shipped.** The 96-length table lives in a docstring; every shipped cell is one `setenv` plus one execute (`grep -n "range(96)\|for pad\|pad_len"` returns nothing). By the PR's own method rule these are the shape it warns against. Parametrising a coarse residue class for the load-bearing cells makes the offline measurement a regression barrier.

**LOW:** the pre-commit eager hook runs conformance only while the CI job also runs `check_examples_run.py` under the knob. `scratch/eager_gc_repro.md` (199 lines) is committed to the branch.

**Verified:** the CI job sets the knob on both steps and fails loudly (`FAIL (run): ch09_nested_builtin_index`, #1382, red by design until its fix lands ahead of this PR); at head 43/43 runnable examples are trap-free under the knob and 243/244 conformance programs pass. Totals: 12,161 passed / 173 skipped; TESTING.md 12,360 across 184 files; `check_doc_counts` agrees (37 hooks, 10 CI jobs); header "Twenty" == 20 == 20 at this base — to be re-derived after the rebase onto the current tip. No overlap with the #1382 fix: this PR touches `vera/wasm/{context,data,helpers,inference,calls_containers,calls_handlers}.py`; the collector, shadow-stack and region-layout constants live in `vera/codegen/assembly.py`, untouched here. Not yet independently reproduced: the 143-mover instruction-multiset accounting and the #1255 live-root counts — to be run against the rebased head.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -49,6 +49,13 @@ repos:
         pass_filenames: false
         files: '(tests/conformance/.*|vera/.*\.py$|vera/grammar\.lark$)'
 
+      - id: conformance-suite-eager-gc
+        name: conformance suite (eager GC)
+        entry: env VERA_EAGER_GC=1 .venv/bin/python scripts/check_conformance.py
+        language: system
+        pass_filenames: false
+        files: '(tests/conformance/.*|vera/.*\.py$|vera/grammar\.lark$)'
```
[the remark this answers is not in the document]
Comment (aallan) [reply .pre-commit-config.yaml:57]: @coderabbitai Already in as of the previous round and carried through the rebase — `vera-examples-run-eager-gc` sits beside the conformance hook in `.pre-commit-config.yaml`, and the documented totals are 39 hooks / 37 commit-stage. Landing SHA 5e211538.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -89,7 +89,7 @@ dependencies. CI enforces that `uv.lock` stays current.
...
 
-The repository configures 36 hooks across two stages: 34 run at the commit stage (after `pre-commit install`), and 2 (`check-changelog-updated` and `uv-lock-check`, described below) run at the push stage (after `pre-commit install --hook-type pre-push`). Most commit-stage hooks have per-hook `files:` / `types:` filters — the `python` type-check only runs when Python files are staged; `check_readme_examples.py` only runs when `README.md` or Vera sources change, etc. A plain-text commit touching only one markdown file triggers a small subset; a compiler-level commit triggers most of them.
+The repository configures 37 hooks across two stages: 35 run at the commit stage (after `pre-commit install`), and 2 (`check-changelog-updated` and `uv-lock-check`, described below) run at the push stage (after `pre-commit install --hook-type pre-push`). Most commit-stage hooks have per-hook `files:` / `types:` filters — the `python` type-check only runs when Python files are staged; `check_readme_examples.py` only runs when `README.md` or Vera sources change, etc. A plain-text commit touching only one markdown file triggers a small subset; a compiler-level commit triggers most of them.
```
[the remark this answers is not in the document]
Comment (aallan) [reply CONTRIBUTING.md]: @coderabbitai Fixed in 5e211538. The duplicate at line 96 read "35 total, of which 34 are gated"; both halves now read 37 / 36, matching line 92 and the `check_doc_counts.py` oracle, which gates this number.

File: scratch/eager_gc_repro.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+(`release/v0.2.0` with #1370 merged).
+
+Everything here needs `VERA_EAGER_GC=1` — a collection at every `$alloc`.  The
+normal gate is green at head: suite 12,152, conformance 244/244, examples
+43/43 (also clean under eager GC).
+
+## The failure
+
+    cd /Users/aa/Code/vera-d1
+    VERA_EAGER_GC=1 PYTHONPATH=$PWD ./.venv/bin/python scripts/check_conformance.py
```
[the remark this answers is not in the document]
Comment (aallan) [reply scratch/eager_gc_repro.md]: @coderabbitai Declining as moot — `scratch/eager_gc_repro.md` was removed in the previous round. It was working notes for the parallel GC hunt on #1382, and the content now lives in that issue and in this PR body, so there is nothing left to make repository-relative.

File: tests/test_codegen_match_shadow_scope_1322.py [hunk 4]
```diff
@@ -294,8 +294,10 @@ class TestMatchCostsNoShadowRoots1322:
             _int_valued_matches,
             _adt_field_matches,
             _allocating_scrutinee_matches,
+            _pointer_result_matches,
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_match_shadow_scope_1322.py:302]: @coderabbitai Fixed in 5e211538, both halves. `_pointer_result_matches` was written against #1322 as the residual and now records the #1371 flip: 4/7/13 roots per frame before #1322, 2/3/5 after it, flat at 1 here. The differential docstring now says four shapes and enumerates the fourth root.

File: tests/test_codegen_root_lifetime_1371.py [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+def _roots_per_frame(make_source) -> int:
+    """Shadow roots one recursive frame holds, from its trap depth.
+
+    The physical quantity: a frame holding *r* roots traps at 4 096 / *r*
+    levels.  The depth also carries the outermost frame's own constant,
+    which does not repeat and which comparing raw depths would read as
+    slope; rounding the ratio drops it and keeps the slope, which is what
+    the defect moved.
+    """
+    return round(_SHADOW_ROOTS / _max_depth(make_source))
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_root_lifetime_1371.py]: @coderabbitai Both fixed in 5e211538. `_compile_ok` moved outside the guard in `_runs`, so a program that stops compiling raises rather than being counted as "costs more than one root per frame" — a measurement failure disguised as a measurement, which is the more valuable half. `_roots_per_frame` is memoised behind an `lru_cache`, removing the redundant binary searches.

File: tests/test_codegen_root_lifetime_1371.py [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+    why the rule reads its shape from `_stack_shape_of` rather than testing
+    for "one i32": a rule keyed on the scalar case roots three of the four and
+    silently misses the String.
+
+    Every claim here is a PAD SWEEP, never a single layout.  #1382 showed a
+    single-layout observation can inverate its own meaning, and two of these
+    cells were green at one layout while being wrong at all of them — the
+    earlier `decimal_to_string` cell passed only because it read the LENGTH,
+    which survives a use-after-free that corrupts the bytes.  Measured over
+    pad lengths 0..95, at this PR's base and at its head:
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_root_lifetime_1371.py]: @coderabbitai Fixed in the previous round and carried through the rebase: `_PADS` is eight lengths covering every residue mod 8, and each load-bearing cell takes a `pad` axis via `_with_pad`. The docstring now also states that the 0..95 sweeps were the offline measurement that selected these cells.

File: vera/README.md [hunk 7]
```diff
@@ -96,7 +96,7 @@ execute(compile_result, ...)    # → run WASM via wasmtime
 | `smt.py` | 3,289 | Verify | Z3 translation layer; reads each callee's contract in the module that declared it (`_callee_contract_scope`), swapping the naming env its slots render against and the registry its bare-name calls resolve in as one `CalleeScope` (#1208, #1225) | `SmtContext`, `SlotEnv`, `CalleeScope` |
 | `verifier.py` | 9,446 | Verify | Contract verification; owns the per-module registries every rendering goes through — an imported callee's contract and an imported generic's clone are named, resolved, and quoted in the module that **declared** them (#1208, #1220, #1225) | `verify()` |
 | `wasm/` | 27,524 | Compile | WASM translation layer (package) | `WasmContext`, `WasmSlotEnv`, `StringPool` |
-| ` ├ context.py` | 1,292 | | Composed WasmContext, expression dispatcher, block translation | |
+| ` ├ context.py` | 1,613 | | Composed WasmContext, expression dispatcher, block translation | |
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/README.md]: @coderabbitai Fixed in 5e211538. Rebasing onto the stacked head re-derived every count from the oracle: 12,712 tests across 191 files, 247 conformance programs. `check_doc_counts.py` is clean.

File: vera/wasm/calls_containers.py [hunk 8]
```diff
@@ -9,8 +9,9 @@
 return the wrapper pointer directly — the wrapper's bucket array (in
 ``vera/codegen/api.py``) IS the map / set, so there is no host-side
 store and no per-call wrap/unwrap.  Copy-on-write ops (insert / add /
-remove / new) return a fresh wrapper that the call-site shadow-roots
-(``_emit_root_result``).  Decimal is the one type that keeps the
+remove / new) return a fresh wrapper, rooted where it LANDS by the
+call-result rule in ``WasmContext._root_landed_value`` (#1379) rather
+than by a per-site push here.  Decimal is the one type that keeps the
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_containers.py:14]: @coderabbitai Fixed in 5e211538 — the three stale comments now name `_root_landed_value` instead of the deleted per-site push. Your reading of the two paths is right and is what the mutation confirms: with the rule disabled and the per-site pushes gone, three Map cells die with a bus error.

File: vera/wasm/context.py [hunk 9]
```diff
@@ -846,10 +855,318 @@ def _mark_byte_write_value(
...
+        if not self.needs_alloc:
+            # A push was emitted, so `$gc_sp` must have been declared.  Every
+            # producer sets the flag beside its push; a new one that forgets
+            # would emit a module referencing an undeclared global, and this
+            # says so at the seam rather than at WAT assembly.
+            raise CodegenInvariantError(
+                f"{type(expr).__name__} lowering emitted a GC shadow push "
+                "without setting needs_alloc, so `$gc_sp` would not be "
+                "declared"
+            )
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/context.py:936]: @coderabbitai Correct, and it generalises — fixed in 5e211538. Both helpers now set `needs_alloc` beside their pushes, and a sweep found two more of the same shape (`_emit_wrap_handle`, `_show_hash_gc_frame`), four in total on top of the five #1376 fixed. Nine sites over two rounds is not a discipline anyone holds by hand, so `TestEveryPushSiteDeclaresTheGlobal1376` now walks every `gc_shadow_push` site in `vera/` and fails on any that never assigns the flag. Three are exempt because they are guarded by construction (two emit inside `if ctx.needs_alloc:`; the scoping wrapper only runs when a push already happened and raises otherwise), and a second test fails if an exemption stops describing a real site.

File: vera/wasm/context.py [hunk 10]
```diff
@@ -1013,11 +1330,46 @@ def translate_expr(
...
+        value is dropped — and every one of them is dead the moment the
+        statement ends.  What outlives a statement is exactly what it BOUND,
+        so the statement restores ``$gc_sp`` to where it stood on entry and
+        then roots its new bindings, read from the environment delta.
+
+        That delta is why the producers no longer root their own bindings:
+        one rule decides which of a statement's bindings are heap pointers
+        and roots each exactly once, where before `let`, pair-`let` and
+        `let`-destructure each pushed for themselves — and a scoped restore
+        would have had to know, per producer, what to put back.
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/context.py]: @coderabbitai Correct — fixed in 5e211538. Your account of the runtime behaviour matches what I measured, so the code is unchanged and the invariant is restated: `let` and pair-`let` had their producer pushes removed, `let`-destructure keeps its own, those land inside the statement instructions and are reclaimed by the restore, and the delta then plants the surviving root — one live root per binding either way. The comment now says which of the three it is talking about, since that is what a future reader would use to decide whether a push is redundant.

File: vera/wasm/inference.py [hunk 11]
```diff
@@ -435,9 +435,22 @@ def _infer_fncall_wasm_type(self, expr: ast.FnCall) -> str | None:
...
+        # absorbed, including another module's `fn show` that this
+        # declaration's lexical scope does not contain.  Asked by membership,
+        # this arm answered "the user's" for `show(42)` in a file that merely
+        # IMPORTS such a module, and typed a String result as `i64` — while
+        # `_translate_call`, narrowed to `_scoped_fns` by #1299, correctly
+        # lowered the ability operation.  The two consultors disagreeing is
+        # what #1299 fixed at the lowering and left here.
+        if (expr.name in ("show", "hash")
+                and not self._bare_call_denotes_op(expr.name)
+                and expr.name in self._fn_ret_types):
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/inference.py:454]: @coderabbitai Correct — fixed in 5e211538. `_infer_fncall_vera_type` now asks `_bare_call_denotes_op` before the `show` / `hash` special cases, so it agrees with `_infer_fncall_wasm_type` and `_translate_call`. This is the same seam #1299 fixed at the lowering; I had aligned the width consultor in this PR and missed its Vera-type twin, which is exactly the flat-registry-by-membership failure the width one had.

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -49,6 +49,20 @@ repos:
...
+
+      - id: vera-examples-run-eager-gc
+        name: vera examples run (eager GC)
+        entry: env VERA_EAGER_GC=1 .venv/bin/python scripts/check_examples_run.py
+        language: system
+        pass_filenames: false
+        files: '(examples/.*|vera/.*\.py$)'
```
Comment (aallan) [inline .pre-commit-config.yaml:64]: @coderabbitai Declining — the premise no longer holds. #1382 was fixed by the collector change that landed ahead of this branch, and rebasing onto that head makes the lane green end to end: 247/247 conformance programs and every runnable example under `VERA_EAGER_GC=1`. The hook blocks nothing, and the lane is what stops #1379 and #1382 re-entering silently, since both were green in the ordinary run and failing under the knob.

File: tests/test_codegen_root_lifetime_1371.py [hunk 13]
```diff
@@ -0,0 +1,906 @@
...
+    into a non-zero return code that an ordinary assertion can name.
+    """
+    src = tmp_path / "cell.vera"
+    src.write_text(source, encoding="utf-8")
+    proc = subprocess.run(
+        [sys.executable, "-m", "vera.cli", "run", str(src)],
+        check=False,  # a crash IS the observation; the caller asserts on rc
+        capture_output=True, text=True, encoding="utf-8",
+        env={"PATH": "/usr/bin:/bin", "VERA_EAGER_GC": "1",
+             "PYTHONPATH": str(Path(__file__).resolve().parents[1])},
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_root_lifetime_1371.py]: @coderabbitai Fixed in 5e211538 — the child now extends `os.environ` and overrides only `VERA_EAGER_GC` and `PYTHONPATH`. The Windows failure mode you name is the sharp one: a missing SYSTEMROOT kills Python before Vera runs, and the cell would then report a launch failure as an unrooted wrapper.

File: tests/test_codegen_root_lifetime_1371.py [hunk 14]
```diff
@@ -0,0 +1,906 @@
...
+            continue
+        # A pair pops its length first, so the pointer half is set last.
+        ptr_local = captured[-1]
+        text = "\n".join(window)
+        pushed = {int(x) for x in _PUSH_OF.findall(text)}
+        if ptr_local in pushed:
+            continue
+        if "call $register_wrapper" in text and pushed:
+            continue
+        unrooted.append((name, f"local {ptr_local} is never pushed"))
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_root_lifetime_1371.py:897]: @coderabbitai Correct, and it was the exact property the docstring claimed — fixed in 5e211538. The window now truncates at the first allocation that could collect the result. One wrinkle worth recording: for the #573 wrap/unwrap builtins the `$alloc` immediately after the call IS the wrapping, so truncating there reported all twelve Decimal sites as unrooted; the boundary is the first allocation after `$register_wrapper`. Re-validated by mutation — still red with the rule disabled.

**Comment by aallan** [comment]
@coderabbitai review

File: vera/README.md [hunk 15]
```diff
@@ -754,7 +754,11 @@ The `ERROR_CODES` dict in `errors.py` maps every code to a short description (16
...
+<<<<<<< HEAD
+Testing spans a **pytest suite** of 12,712 tests across 191 files: compiler-internals unit tests plus a **conformance suite** (247 programs in `tests/conformance/` validating every language feature against the spec) and **example programs** (43 end-to-end demos). The conformance suite is the definitive specification artifact; most programs target a single feature, though some (slot references, match, contracts) span several, and each serves as a minimal working example.
+=======
+Testing spans a **pytest suite** of 12,433 tests across 184 files: compiler-internals unit tests plus a **conformance suite** (244 programs in `tests/conformance/` validating every language feature against the spec) and **example programs** (43 end-to-end demos). The conformance suite is the definitive specification artifact; most programs target a single feature, though some (slot references, match, contracts) span several, and each serves as a minimal working example.
+>>>>>>> 67595863 (Carry the landing half on evidence that cannot crash, and sweep the pads that ship)
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/README.md]: @coderabbitai Fixed at the current head d51ef55a: the rebase onto the release tip rebuilt `vera/README.md` from the tip's version plus this PR's own diff with the markers stripped — `git grep -nE '^(<<<<<<<|=======|>>>>>>>)( |$)'` over the whole tree at d51ef55a returns nothing — and `scripts/check_doc_counts.py` reports every documentation count consistent there (13,087 tests across 197 files, 248 conformance programs).

File: tests/test_codegen_root_lifetime_1371.py [hunk 16]
```diff
@@ -0,0 +1,1027 @@
...
+    """Shadow roots one recursive frame holds, from its trap depth.
+
+    The physical quantity: a frame holding *r* roots traps at 4 096 / *r*
+    levels.  The depth also carries the outermost frame's own constant,
+    which does not repeat and which comparing raw depths would read as
+    slope; rounding the ratio drops it and keeps the slope, which is what
+    the defect moved.
+    """
+    return _roots_per_frame_cached(
+        getattr(make_source, "__cache_key__", make_source), make_source)
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_root_lifetime_1371.py:170]: @coderabbitai Accepted as optional: the closure-per-measurement key costs 18 extra measurements in a test that already runs in seconds and keeps no state across cells. Left as is for this PR; a factory-and-k key is a fine follow-up if the file grows.

File: .pre-commit-config.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (aallan) [reply .pre-commit-config.yaml:64]: @coderabbitai Valid, and deferred to the v0.2.0 release-machinery PR on this same line, which is where the hook table is being consolidated: the eager `examples-run` hook will take the same `files` filter as the normal one (grammar, TESTING.md and the checker script included). The CI job already runs both lanes on every push, so the gap is local-only.

File: vera/wasm/calls_arrays.py [hunk 17]
```diff
@@ -1282,6 +1282,11 @@ def _translate_array_fold(
         instructions.extend(arr_instrs)
         instructions.append(f"local.set {arr_len}")
         instructions.append(f"local.set {arr_ptr}")
+        # `needs_alloc` beside the push (#1376/#1379): the flag is what
+        # declares `$gc_sp`, so a push without it emits a reference to
+        # an undeclared global whenever nothing else in the module
+        # allocates — and `_scope_shadow_roots` refuses it outright.
+        self.needs_alloc = True
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_arrays.py:1289]: @coderabbitai Measured at d51ef55a rather than argued. The layout half of the finding is right: for an anonymous-function callback the closure expression's own restore-and-re-root leaves one slot between the accumulator's root and `fn_tmp`, so at loop entry the shadow stack reads [arr, acc, closure-re-root, fn_tmp] and `gc_sp - 8` addresses the closure's re-root slot, not the accumulator's (the emitted WAT for a String-accumulating fold shows exactly that sequence). The consequence differs from the finding, though: the write-back stores the NEW accumulator pointer into that re-root slot every iteration, and the closure remains rooted through its own `fn_tmp` slot, so the current accumulator is always rooted and the only effect is retention of the initial accumulator in its original slot. Empirically: a six-element String fold with a content assertion and an Option<Int> fold that allocates per hit both return the right value under VERA_EAGER_GC=1 at all 40 data-section layouts tried (0/40 failures on each branch; identical at the release tip). So this is sound but fragile, and the remedy stands: capture the accumulator slot's address at push time and store through it in both branches. That change lands in the v0.2.0 release-machinery PR on this same line, with the comment corrected, rather than restarting this PR's wall.

File: vera/wasm/context.py [hunk 18]
```diff
@@ -846,10 +855,318 @@ def _mark_byte_write_value(
...
+        stack shape is unrecoverable.  Guessing it wrong is not a missed
+        reclamation but invalid WASM (a ``local.set`` against an empty
+        operand stack), so an unknown shape keeps the pre-#1371 behaviour:
+        the root outlives its value, which is wasteful and sound.
+        """
+        if not contains_shadow_push(instructions):
+            # Nothing to reclaim — but a CALL still LANDS a value that no
+            # root holds yet, and the next sibling to be evaluated may
+            # allocate over it.
+            return self._root_landed_value(expr, instructions)
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/context.py:926]: @coderabbitai Accepted as a follow-up probe rather than a defect: no cell in this PR targets a pointer returned through `return_call` specifically. Coverage that does bear on it at d51ef55a: the eager-GC lane is green for all 248 conformance programs and every runnable example (the tail-call programs among them included), and the reviewer's swept attacks covered recursion that re-binds a heap value at every level. A dedicated tail-position cell (a tail-recursive function returning a heap value, allocating after the return lands, under VERA_EAGER_GC=1 across a pad sweep) is queued for the v0.2.0 release-machinery PR on this line.
